#!/usr/bin/env python
PACKAGE = "duo3d_driver"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

# DUO camera parameters
#       Name                  Type   Level Description                     Def Min Max
gen.add("gain",             double_t,  0, "Image gain value",               0,  0, 100)
gen.add("exposure",         double_t,  0, "Image exposure value",           50, 0, 100)
gen.add("auto_exposure",      bool_t,  0, "Image auto exposure value",      False)
gen.add("camera_swap",        bool_t,  0, "Image swap",                     False)
gen.add("horizontal_flip",    bool_t,  0, "Horizontally image flip",        False)
gen.add("vertical_flip",      bool_t,  0, "Vertically image flip",          False)
gen.add("led",	            double_t,  0, "LED brightness value",           10, 0, 100)
accel_range_enum = gen.enum([ gen.const("2g",  int_t, 0, "Accelerometer full scale range +/- 2g"),
                              gen.const("4g",  int_t, 1, "Accelerometer full scale range +/- 4g"),
                              gen.const("8g",  int_t, 2, "Accelerometer full scale range +/- 8g"),
                              gen.const("16g", int_t, 3, "Accelerometer full scale range +/- 16g")
                            ], "Enum to set Accelerometer Range")
gen.add("accel_range",	       int_t,  0, "Accelerometer Range",            0,  0, 3, edit_method = accel_range_enum)
gyro_range_enum = gen.enum([ gen.const("250deg_s",  int_t, 0, "Gyroscope full scale range 250 deg/s"),
                             gen.const("500deg_s",  int_t, 1, "Gyroscope full scale range 500 deg/s"),
                             gen.const("1000deg_s", int_t, 2, "Gyroscope full scale range 1000 deg/s"),
                             gen.const("2000deg_s", int_t, 3, "Gyroscope full scale range 2000 deg/s")
                           ], "Enum to set Gyroscope Range")
gen.add("gyro_range",	       int_t,  0, "Gyroscope Range",                0,   0, 3, edit_method = gyro_range_enum)
gen.add("imu_rate",	        double_t,  0, "IMU Sampling Rate",            100,  50, 500)

# Dense3D parameters
#       Name                   Type Level Description                      Def Min Max
proc_mode_enum = gen.enum([ gen.const("BM",   int_t, 0, "Block Matching"),
                            gen.const("SGBM", int_t, 1, "Semiglobal Matching") 
                          ], "Enum to set Processing Mode")
gen.add("processing_mode",     int_t, 0, "Processing Mode",                 0,  0, 1, edit_method = proc_mode_enum)
image_scale_enum = gen.enum([ gen.const("NoScaling", int_t, 0, "No Image Scaling"),
                              gen.const("ScaleX",    int_t, 1, "Scale along X"),
                              gen.const("ScaleY",    int_t, 2, "Scale along Y"),
                              gen.const("ScaleXY",   int_t, 3, "Scale along X & Y")
                            ], "Enum to set Image Scaling Mode")
gen.add("image_scale",         int_t, 0, "Image Scaling Mode",              3,  0, 3, edit_method = image_scale_enum)
gen.add("pre_filter_cap",      int_t, 0, "Pre-filter cap (dense3D)", 		28, 1, 63)
gen.add("num_disparities",     int_t, 0, "Number of disparities (dense3D)", 4,  2, 16)
gen.add("sad_window_size",     int_t, 0, "SAD Window Size (dense3D)", 		6,  2, 10)
gen.add("uniqueness_ratio",	   int_t, 0, "Uniqueness Ratio (dense3D)",      27, 1, 100)
gen.add("speckle_window_size", int_t, 0, "Speckle Window Size (dense3D)",   52, 0, 256)
gen.add("speckle_range",       int_t, 0, "Speckle Range (dense3D)",         14, 0, 32)

exit(gen.generate(PACKAGE, "duo3d_driver", "Duo3D"))
